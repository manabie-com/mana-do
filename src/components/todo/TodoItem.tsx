import React, { useEffect, useRef, useState } from 'react';
import { TodoStatus, Todo } from '../../models/todo';
import useClickOutside from '../../hooks/useClickOutside';

interface TodoItemProps {
  todo: Todo;
  onDelete?: (id: string) => void;
  onComplete?: (id: string, isComplete: boolean) => void;
  onUpdateContent?: (id: string, content: string) => void;
}

const TodoItem = (props: TodoItemProps) => {
  const [isUpdateContent, setIsUpdateContent] = useState<boolean>(false);
  const [editingContent, setEditingContent] = useState<string>();
  const itemRef = useRef<HTMLDivElement>(null);
  const editingContentRef = useRef<HTMLTextAreaElement>(null);

  useClickOutside(itemRef, () => {
    if (isUpdateContent === true) {
      setIsUpdateContent(false);
    }
  });

  useEffect(() => {
    if (isUpdateContent) {
      const length = editingContentRef.current?.value.length || null;
      editingContentRef.current?.setSelectionRange(length, length);
      editingContentRef.current?.focus();
    }
  }, [isUpdateContent]);

  const onEnableUpdateContent = (e: React.MouseEvent<HTMLParagraphElement>) => {
    // double click
    if (e.detail === 2) {
      setEditingContent(props.todo.content);
      setIsUpdateContent(true);
    }
  };

  const onUpdateTodoContent = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    if (e.key === 'Enter' && editingContent?.trim()) {
      props.onUpdateContent &&
        props.onUpdateContent(props.todo.id, editingContent.trim());
      setIsUpdateContent(false);
    }
  };

  return (
    <div ref={itemRef} className="item">
      {isUpdateContent ? (
        <textarea
          ref={editingContentRef}
          className="item__content item__content-edit"
          value={editingContent}
          onChange={(e) => setEditingContent(e.target.value)}
          onKeyDown={onUpdateTodoContent}
        />
      ) : (
        <p
          className="item__content item__content-noedit"
          onClick={onEnableUpdateContent}
        >
          {props.todo.content}
        </p>
      )}

      <div className="item__actions">
        <button
          className="item__delete"
          onClick={() => props.onDelete && props.onDelete(props.todo.id)}
        >
          X
        </button>
        <div className="action__checkbox">
          <input
            id={`completed-${props.todo.id}`}
            type="checkbox"
            checked={TodoStatus.COMPLETED === props.todo.status}
            onChange={(e) =>
              props.onComplete &&
              props.onComplete(props.todo.id, e.target.checked)
            }
          />
          <label htmlFor={`completed-${props.todo.id}`}>Completed</label>
        </div>
      </div>
    </div>
  );
};

export default TodoItem;
