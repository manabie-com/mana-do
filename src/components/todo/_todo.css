
button {
  outline: none;
  border: none;
  box-shadow: 2px 0 2px currentColor;
  border-radius: 4px;
  min-height: 32px;
  min-width: 80px;
  padding: 4px 8px;
  color: steelblue;
  font-weight: 600;
}

button:hover {
  opacity: 0.85;
  border: 1px solid rgba(70,130,180, 0.7);
}

input[type="checkbox"] {
  width: 24px;
  height: 24px;
  box-shadow: none;
  border: none;
  outline: none;
  cursor: pointer;
  width: 1.5em;
  height: 1.5em;
}

input[type="checkbox"]:focus {
  box-shadow: none;
  border: none;
  outline: none;
}

input {
  min-height: 36px;
  border: none;
  outline: none;
  padding: 0 12px;
  box-shadow: 2px 0 4px rgba(0,0,0, 0.2);
  border-radius: 4px;
}

input:focus {
  box-shadow: 1px 0 9px rgba(0,0,0, 0.25);
}

.ToDo__Wrapper {
  background: url('./../../assets/images/bg_ocean.jpeg');
  background-repeat: no-repeat;
  background-size: cover;
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
}

.ToDo__container {
  border: 1px solid rgba(0,0,0, 0.4);
  border-radius: 8px;
  max-width: 700px;
  min-width: 350px;
  height: 60vh;
  width: 50vw;
  margin-top: 5rem;
  padding: 24px;
  box-shadow: 2px 2px 1px rgba(0,0,0, 0.09),
              3px 2px 3px rgba(0,0,0, 0.05);
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  background-color: rgba(70,130,180, 0.25);
  animation: form-appear 2s;
  opacity: 1;
}

@keyframes form-appear {
  from {
    opacity: 0;
    transform: translateY(-70px);
  }
  to {
    opacity: 1;
  }
}

.Todo__creation {
  display: flex;
  width: 100%;
}

input.TodoItem__input {
  width: 80%;
  background-color: rgba(255,255,255, 0.1);
  color: white;
  transition: 1s;
}

input.TodoItem__input:focus {
  box-shadow: 2px 0 4px rgba(213, 250, 1, 0.7);
}

input.TodoItem__input:hover {
  box-shadow: 2px 0 4px rgba(250, 113, 1, 0.7);
}

.Todo__input {
  flex: 1 1;
  background-color: rgba(70,130,180, 0.7);
  color: white;
  transition: 1s;
}

.Todo__input::placeholder {
  color: rgba(240, 247, 179, 0.7);
}

.Todo__input:focus {
  box-shadow: 2px 0 4px rgba(213, 250, 1, 0.7);
}

.Todo__input:hover {
  box-shadow: 2px 0 4px rgba(250, 113, 1, 0.7);
}

.ToDo__list {
  display: flex;
  flex-direction: column;
  margin-top: 1.5rem;
  width: 100%;
  height: 100%;
  overflow-y: auto;
  border-right: solid 1px rgba(127, 255, 212, 0.9);
  border-left: solid 1px rgba(127, 255, 212, 0.9);
  border-bottom: solid 2px rgba(127, 255, 212, 0.9);
  border-bottom-left-radius: 20px;
  border-bottom-right-radius: 20px;
  background-color: rgba(127, 255, 212, 0.2);
  box-shadow: 4px 0 4px currentColor;
}

.ToDo__item {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 4px 2px;
  padding: 2px 4px;
  border: 1px solid rgba(70,130,180, 0.4);
  border-radius: 15px;
  background-color: rgba(70,130,180, 0.4);
  color: white;
  animation: item-appear 0.4s;
  cursor: pointer;
  transition: 0.5s;
  font-size: 16px;
  min-height: 2.5em;
}

.ToDo__item:hover {
  background-color: rgba(240, 247, 181, 0.3);
}

@keyframes item-appear {
  from {
    opacity: 0;
    transform: translateY(30px);
  }
  to {
    opacity: 1;
  }
}

.ToDo__item > span {
  flex: 1 1;
  text-align: left;
  margin-left: 8px;
  margin-right: 8px;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  word-break: break-word;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 5px;
  font-size: 16px;
  line-height: 24px;
}

.Todo__content {
  flex: 1 1;
}

.Todo__delete {
  outline: none;
  border: none;
  width: 32px;
  height: 32px;
  min-width: auto;
  min-height: auto;
  box-shadow: none;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  background-color: rgba(255, 0, 0, 0.6);
  color: white;
}

.Todo__action, .Todo__delete {
  width: 24px;
  height: 24px;
  flex-shrink: 0;
}

.Todo__toolbar {
  display: flex;
  justify-content: space-between;
  margin-top: 24px;
}

.Todo__tabs {
  display: flex;
  justify-content: center;
}

.Todo__tabs > *:not(:last-child) {
  /* margin-right: 8px; */
  border-right: 2px solid rgba(70,130,180, 0.4);
}

.Action__btn {
  cursor: pointer;
  background-color: aquamarine;
  transition-duration: 0.5s;
}

.Action__btn:hover {
  transform: translateY(-5px);
}

.Action__btn.active {
  background-color: rgba(70,130,180, 0.8);
  color: white;
}

.Action__btn:disabled {
  cursor: not-allowed;
}

.use-tooltip {
  position: relative;
  display: inline;
}
  
.use-tooltip:hover .tooltip {
  display: block;
  position: fixed;
  z-index: 3;
  left: 50%;
  top: -20px;
  transform: translateX(-50%);
  min-width: 80vw;
  text-align: left;
  display: -webkit-box;
  -webkit-line-clamp: 5;
  word-break: break-word;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0.5em;
  font-size: 16px;
  line-height: 24px;
  color: steelblue;
  font-weight: 500;
}
  
.tooltip {
  display: none;
  color: white;
  background-color: rgb(240, 247, 181, 0.7);
  border-radius: 20px;
  margin: 2em;
}

@media screen and (min-width: 1201px) {
}
@media screen and (max-width: 1200px) {
}
@media screen and (max-width: 1024px) {
  .ToDo__container {
    width: 70vw;
  }
}
@media screen and (max-width: 768px) {
  .ToDo__container {
    width: 100vw;
  }
}
@media screen and (max-width: 480px) {
}