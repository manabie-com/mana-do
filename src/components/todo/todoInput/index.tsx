import React from "react";
import { ITodoProps } from "./ITodoProps";
import "./todoInput.scss";

const TodoInput = (props: ITodoProps): JSX.Element => {
  return (
    <div className="todo-input">
      <input {...props} className="todo-input__input" />
    </div>
  );
};

export default TodoInput;
