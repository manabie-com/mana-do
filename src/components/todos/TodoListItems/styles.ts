import styled from "styled-components";
import { TODO_ITEM_HEIGHT } from "../../../utils/constants";

export const ListItemWrapper = styled.article`
  display: flex;
  width: 100%;
  min-height: ${TODO_ITEM_HEIGHT}px;
  padding: 12px 16px;
  color: var(--font-color);
  background-color: var(--listbg);
  vertical-align: middle;
  border-bottom: 1px solid var(--softbg);

  :first-of-type {
    border-radius: 5px 5px 0px 0px;
  }

  .title {
    width: 5000px;
    margin-bottom: 0;
    color: var(--font-color);
    transition: var(--transition);
    text-decoration: none;
    align-self: center;
    overflow: hidden;
    height: 20px;
    text-align: left;
  }

  .title-completed {
    width: 5000px;
    margin-bottom: 0;
    color: var(--font-color);
    transition: var(--transition);
    text-decoration: line-through;
    align-self: center;
    height: 20px;
    vertical-align: middle;
    overflow: hidden;
  }

  .completed-btn,
  .uncompleted-btn {
    width: 20px;
    align-items: center;
    cursor: pointer;
    transition: var(--transition);
    border-radius: 50%;
    margin-right: 1rem;
    height: 20px;
    vertical-align: middle;
  }

  .uncompleted-btn {
    border: 2px solid var(--softbg);
    background: none;
  }

  .completed-btn {
    border: none;
    background-size: cover;
    background: url(./assets/icons/icon-check.svg) no-repeat, linear-gradient(hsl(192, 100%, 67%), hsl(280, 87%, 65%));
    background-position-x: center;
    background-position-y: center;
  }

  .completed-btn:hover,
  .uncompleted-btn:hover {
    border-color: var(--font-color);
  }

  .completed-btn:focus,
  .uncompleted-btn:focus,
  .completed-btn:active,
  .uncompleted-btn:active {
    outline: none;
  }

  .delete {
    display: flex;
    justify-content: flex-end;
    align-self: center;
    margin-left: 14px;
  }

  .delete-btn {
    color: var(--softbg);
    align-items: center;
    cursor: pointer;
    border: none;
    font-size: 20px;
    background: none;
    height: 15px;
    vertical-align: middle;

    :focus {
      outline: none;
    }

    svg {
      transition: transform 0.25s, opacity 0.25s;
      opacity: 0.5;

      path {
        fill: var(--font-color);
      }
    }

    :hover {
      svg {
        transform: rotate(270deg);
        opacity: 1;

        path {
          fill: red;
        }
      }
    }
  }

  .btn-cont {
    align-self: center;
  }

  .text_and_checkbox {
    --text: #414856;
    --check: var(--brightblue);
    --disabled: #c3c8de;
    --border-radius: 10px;
    border-radius: var(--border-radius);
    box-shadow: 0 10px 30px rgba(#414856, 0.05);
    display: grid;
    grid-template-columns: 30px auto;
    align-items: center;
    width: 100%;

    label {
      color: var(--font-color);
      position: relative;
      cursor: pointer;
      display: grid;
      align-items: center;
      width: fit-content;
      transition: color 0.3s ease;
      &::before,
      &::after {
        content: "";
        position: absolute;
      }
      &::before {
        height: 2px;
        width: 8px;
        left: -27px;
        background: var(--check);
        border-radius: 2px;
        transition: background 0.3s ease;
      }
      &:after {
        height: 4px;
        width: 4px;
        top: 8px;
        left: -25px;
        border-radius: 50%;
      }
    }

    .input_live_edit {
      width: 100%;
      height: 24px;
      transition: none;
      border: none;
      outline: none;

      :focus,
      :hover {
        border: none;
      }
    }

    input[type="checkbox"] {
      -webkit-appearance: none;
      -moz-appearance: none;
      position: relative;
      height: 15px;
      width: 15px;
      outline: none;
      border: 0;
      margin: 0 15px 0 0;
      cursor: pointer;
      display: grid;
      align-items: center;
      &::before,
      &::after {
        content: "";
        position: absolute;
        height: 2px;
        top: auto;
        background: var(--check);
        border-radius: 2px;
      }
      &::before {
        width: 0px;
        right: 60%;
        transform-origin: right bottom;
      }
      &::after {
        width: 0px;
        left: 40%;
        transform-origin: left bottom;
      }
      &:checked {
        &::before {
          animation: check-01 0.4s ease forwards;
        }
        &::after {
          animation: check-02 0.4s ease forwards;
        }
        + label {
          color: var(--disabled);
          animation: move 0.3s ease 0.1s forwards;
          &::before {
            background: var(--disabled);
            animation: slice 0.4s ease forwards;
          }
          &::after {
            animation: firework 0.5s ease forwards 0.1s;
          }
        }
      }
    }
  }

  @keyframes move {
    50% {
      padding-left: 8px;
      padding-right: 0px;
    }
    100% {
      padding-right: 4px;
    }
  }
  @keyframes slice {
    60% {
      width: 100%;
      left: 4px;
    }
    100% {
      width: 100%;
      left: -2px;
      padding-left: 0;
    }
  }
  @keyframes check-01 {
    0% {
      width: 4px;
      top: auto;
      transform: rotate(0);
    }
    50% {
      width: 0px;
      top: auto;
      transform: rotate(0);
    }
    51% {
      width: 0px;
      top: 8px;
      transform: rotate(45deg);
    }
    100% {
      width: 5px;
      top: 8px;
      transform: rotate(45deg);
    }
  }
  @keyframes check-02 {
    0% {
      width: 4px;
      top: auto;
      transform: rotate(0);
    }
    50% {
      width: 0px;
      top: auto;
      transform: rotate(0);
    }
    51% {
      width: 0px;
      top: 8px;
      transform: rotate(-45deg);
    }
    100% {
      width: 10px;
      top: 8px;
      transform: rotate(-45deg);
    }
  }
  @keyframes firework {
    0% {
      opacity: 1;
      box-shadow: 0 0 0 -2px #4f29f0, 0 0 0 -2px #4f29f0, 0 0 0 -2px #4f29f0, 0 0 0 -2px #4f29f0, 0 0 0 -2px #4f29f0,
        0 0 0 -2px #4f29f0;
    }
    30% {
      opacity: 1;
    }
    100% {
      opacity: 0;
      box-shadow: 0 -15px 0 0px #4f29f0, 14px -8px 0 0px #4f29f0, 14px 8px 0 0px #4f29f0, 0 15px 0 0px #4f29f0,
        -14px 8px 0 0px #4f29f0, -14px -8px 0 0px #4f29f0;
    }
  }
`;
