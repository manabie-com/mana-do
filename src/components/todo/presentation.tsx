/**
 * For applying presentation component pattern
 * Split UI process and logic process.
 */
import React from 'react';
import TodoItem from './subPresentations/todoItem';
import { Todo } from 'root/models/todo';

export type TodoProps = {
  inputRef: any,
  onCreateTodo: React.KeyboardEventHandler,
  showTodos: Array<Todo>,
  onUpdateTodoStatus: Function,
  deleteTodo: Function,
  todos: Array<Todo>,
  activeTodos: Number,
  onToggleAllTodo: React.ChangeEventHandler,
  handleShowingAll: React.MouseEventHandler,
  handleShowingActive: React.MouseEventHandler,
  handleShowingCompleted: React.MouseEventHandler,
  onDeleteAllTodo: React.MouseEventHandler,
}

const text = {
  /** help to concentrate all text, easy to edit text or apply multi-language feature */
  x: 'X',
  all: 'All',
  active: 'Active',
  completed: 'Completed',
  clearAllTodos: 'Clear all todos',
  whatNeedToBeDone: 'What need to be done?'
}


const TodoPresentation = (props: TodoProps) => {
  const {
    inputRef, onCreateTodo, showTodos
    , onUpdateTodoStatus, deleteTodo, todos
    , activeTodos, onToggleAllTodo
    , handleShowingAll, handleShowingActive
    , handleShowingCompleted, onDeleteAllTodo
  } = props

  return (
    <div className='ToDo__container'>
      <div className='Todo__creation'>
        <input
          ref={inputRef}
          className='Todo__input'
          placeholder={text.whatNeedToBeDone}
          onKeyDown={onCreateTodo}
        />
      </div>
      <div className='ToDo__list'>
        {
          showTodos.map((todo: Todo, index: number) => (
            <TodoItem key={`show-todo-item-${index}`}
              todo={todo}
              deleteItemText={text.x}
              deleteTodo={deleteTodo}
              onUpdateTodoStatus={onUpdateTodoStatus}
            />
          ))
        }
      </div>
      <div className='Todo__toolbar'>
        {todos.length > 0 ?
          <input
            type='checkbox'
            checked={activeTodos === 0}
            onChange={onToggleAllTodo}
          /> : <div />
        }
        <div className='Todo__tabs'>
          <button className='Action__btn' onClick={handleShowingAll}>
            {text.all}
          </button>
          <button className='Action__btn' onClick={handleShowingActive}>
            {text.active}
          </button>
          <button className='Action__btn' onClick={handleShowingCompleted}>
            {text.completed}
          </button>
        </div>
        <button className='Action__btn' onClick={onDeleteAllTodo}>
          {text.clearAllTodos}
        </button>
      </div>
    </div>
  );
};

//use shallowly compare complex objects in the props object
export default React.memo(TodoPresentation)