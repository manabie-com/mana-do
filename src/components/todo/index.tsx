import React, { useEffect, useReducer, useRef, useState, useMemo, useCallback } from 'react';
import { RouteComponentProps } from 'react-router-dom';

import reducer, { initialState } from 'root/store/reducer';
import {
  setTodos,
  createTodo,
  deleteTodo,
  toggleAllTodos,
  deleteAllTodos,
  updateTodoStatus,
  updateTodoContent
} from 'root/store/actions';
import Service from 'root/service';
import { TodoStatus } from 'root/models/todo';
import { isTodoCompleted } from 'root/utils';
import TodoPresentation from './presentation';

import './_todo.css'

type EnhanceTodoStatus = TodoStatus | 'ALL';


const TodoComponent = ({ history }: RouteComponentProps) => {
  const [{ todos }, dispatch] = useReducer(reducer, initialState);
  const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    (async () => {
      const resp = await Service.getTodos();
      dispatch(setTodos(resp || []));
    })()
  }, [])

  useEffect(() => {
    if (todos.length > 0) {
      const lastItem = document.getElementById(`todo-item-${todos[todos.length - 1].id}`)
      if (lastItem) setTimeout(() => {
        lastItem.scrollIntoView({ block: 'end',  behavior: 'smooth' })
      }, 401) // delay to wait for css animation work first
    }
  }, [todos, showing])

  const onCreateTodo = useCallback(async (e: React.KeyboardEvent<HTMLInputElement>) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    if (e.key === 'Enter' && inputRef.current) {
      try {
        const resp = await Service.createTodo(inputRef.current.value);
        dispatch(createTodo(resp));
        inputRef.current.value = '';
      } catch (e) {
        if (e.response.status === 401) {
          history.push('/')
        }
      }
    }
  }, [dispatch, history])

  const onUpdateTodoStatus: Function = useCallback(async (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
    try {
      await Service.updateTodoStatus(todoId, e.target.checked);
      dispatch(updateTodoStatus(todoId, e.target.checked));
    } catch (error) {

    }
  }, [dispatch])

  const onToggleAllTodo = useCallback(async (e: React.ChangeEvent<HTMLInputElement>) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.toggleAllTodos(e.target.checked)
      dispatch(toggleAllTodos(e.target.checked))
    } catch (error) {

    }
  }, [dispatch])

  const onDeleteAllTodo = useCallback(async () => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.deleteAllTodo()
      dispatch(deleteAllTodos());
    } catch (error) {

    }
  }, [dispatch])

  const handleDeleteTodo = useCallback(async (todoId: string) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.deleteTodo(todoId)
      dispatch(deleteTodo(todoId));
    } catch (error) {

    }
  }, [dispatch])

  const handleUpdateTodoContent = useCallback(async (todoId: string, content: string) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.updateTodoContent(todoId, content)
      dispatch(updateTodoContent(todoId, content));
    } catch (error) {

    }
  }, [dispatch])

  const handleShowingAll = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing('ALL')
  }, [setShowing])

  const handleShowingActive = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing(TodoStatus.ACTIVE)
  }, [setShowing])

  const handleShowingCompleted = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing(TodoStatus.COMPLETED)
  }, [setShowing])

  const showTodos = useMemo(() => todos.filter((todo) => {
    /**
     * Reduce recalculating showing todos, Does not recalculate showing todos when todos and showing state have not changed
     */
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      case 'ALL':
      default:
        return true;
    }
  }), [showing, todos])

  const activeTodos = useMemo(() => todos.reduce(function (accum, todo) {
    /**
     * Reduce recalculating activeTodos, Does not recalculate active todos when todos has not changed
    */
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0), [todos])

  return (
    <TodoPresentation
      inputRef={inputRef}
      onCreateTodo={onCreateTodo}
      showTodos={showTodos}
      onUpdateTodoStatus={onUpdateTodoStatus}
      deleteTodo={handleDeleteTodo}
      todos={todos}
      activeTodos={activeTodos}
      onToggleAllTodo={onToggleAllTodo}
      handleShowingAll={handleShowingAll}
      handleShowingActive={handleShowingActive}
      handleShowingCompleted={handleShowingCompleted}
      onDeleteAllTodo={onDeleteAllTodo}
      onUpdateTodoContent={handleUpdateTodoContent}
      showing={showing}
    />
  );
};

export default TodoComponent;