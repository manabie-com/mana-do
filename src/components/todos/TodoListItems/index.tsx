import React, { useState } from "react";
import { findTodoById, findTodoIndexById, updateTodoContent } from "../../../utils/todoUtils";
import { isTodoCompleted } from "../../../utils";
import { deleteTodo, updateTodoContent as updateTodoContentAction, updateTodoStatus } from "../../../store/actions";
import { getTodosFromStorage } from "../../../utils/storageUtils";
import { useAppContext } from "../../../AppContext";
import { ListItemWrapper } from "./styles";
import { Draggable } from "react-beautiful-dnd";
import CrossIcon from "../../../assets/icons/CrossIcon";
import { Todo } from "../../../models/todo";

interface TodoListItemsProps {
  id: string;
  index?: number;
}

const TodoListItems = ({ id, index = 0 }: TodoListItemsProps) => {
  const {
    state: { todos },
    dispatch,
  } = useAppContext();
  const [todosValue, setTodosValue] = useState(todos);
  const todo = findTodoById(todos, id) ?? ({} as Todo);
  const [toggle, setToggle] = useState({});
  const foundIndex = findTodoIndexById(todosValue, id);
  const text = todosValue[foundIndex].content;

  const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
    dispatch(updateTodoStatus(todoId, e.target.checked));
  };

  const deleteTodoItem = (todoId: string) => {
    dispatch(deleteTodo(todoId));
  };

  const onDoubleClickItem = (e: React.MouseEvent<HTMLDivElement>) => {
    setToggle((prev) => ({
      ...prev,
      [id]: true,
    }));
  };

  const onLiveEditKeydown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    const target = event.target as HTMLTextAreaElement;
    if (event.key === "Enter" || event.key === "Escape") {
      if (event.key === "Escape") {
        setTodosValue(getTodosFromStorage());
      }

      if (event.key === "Enter") {
        const newTodos = updateTodoContent(todosValue, todo.id, target.value);
        setTodosValue(newTodos);
        dispatch(updateTodoContentAction(todo.id, target.value));
      }

      setToggle((prev) => ({
        ...prev,
        [id]: false,
      }));
    }
  };

  const onChangeLiveEdit = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    const foundIndex = findTodoIndexById(todos, todo.id);
    let newTodos = [...todos];
    newTodos[foundIndex] = {
      ...newTodos[foundIndex],
      content: value,
    };
    setTodosValue(newTodos);
  };

  const onBlurLiveEdit = () => {
    setToggle((prev) => ({
      ...prev,
      [id]: false,
    }));
    setTodosValue(getTodosFromStorage());
  };

  return (
    <Draggable key={id} draggableId={"draggable-" + id} index={index}>
      {(provided, snapshot) => (
        <ListItemWrapper
          key={id}
          className="list-item"
          onDoubleClick={onDoubleClickItem}
          ref={provided.innerRef}
          {...provided.draggableProps}
          {...provided.dragHandleProps}
          data-testid="list-item-wrapper"
        >
          <div className="text_and_checkbox">
            <input
              id={todo.id}
              type="checkbox"
              data-testid="toggle-all"
              checked={isTodoCompleted(todo)}
              onChange={(e) => onUpdateTodoStatus(e, todo.id)}
            />
            {!toggle[id] && <label htmlFor={todo.id}>{text}</label>}
            {toggle[id] && (
              <input
                autoFocus
                type="text"
                className="input_live_edit"
                value={text}
                onChange={onChangeLiveEdit}
                onBlur={onBlurLiveEdit}
                onKeyDown={onLiveEditKeydown}
                data-testid="live-edit-todo-content"
              />
            )}
          </div>

          <div className="delete">
            <button className="delete-btn" data-testid="delete-all-todo-item" onClick={() => deleteTodoItem(id)}>
              <CrossIcon />
            </button>
          </div>
        </ListItemWrapper>
      )}
    </Draggable>
  );
};

export default TodoListItems;
