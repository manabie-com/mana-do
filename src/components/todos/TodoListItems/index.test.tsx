import React from "react";
import renderer from "react-test-renderer";
import TodoListItems from ".";
import * as AppContext from "../../../AppContext";
import { TodoStatus } from "../../../models/todo";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useRef: jest.fn(),
  useState: jest.fn(),
}));

jest.mock("react-beautiful-dnd", () => ({
  Droppable: ({ children }: any) =>
    children(
      {
        draggableProps: {
          style: {},
        },
        innerRef: jest.fn(),
      },
      {}
    ),
  Draggable: ({ children }: any) =>
    children(
      {
        draggableProps: {
          style: {},
        },
        innerRef: jest.fn(),
      },
      {}
    ),
  DragDropContext: ({ children }: any) => children,
}));

const todoId = "hfuwyr";

const mockedTodos = [
  {
    id: todoId,
    content: "todo 1",
    user_id: "firstUser",
    status: TodoStatus.All,
    created_date: "",
  },
  {
    id: "orislaj",
    content: "todo 2",
    user_id: "firstUser",
    status: TodoStatus.All,
    created_date: "",
  },
  {
    id: "jiwuro",
    content: "todo 3",
    user_id: "firstUser",
    status: TodoStatus.Active,
    created_date: "",
  },
  {
    id: "iejals",
    content: "todo 4",
    user_id: "firstUser",
    status: TodoStatus.Active,
    created_date: "",
  },
  {
    id: "jfuqos",
    content: "todo 5",
    user_id: "firstUser",
    status: TodoStatus.Completed,
    created_date: "",
  },
  {
    id: "ahsyel",
    content: "todo 6",
    user_id: "firstUser",
    status: TodoStatus.Completed,
    created_date: "",
  },
];

beforeEach(() => {
  const mockContextValue = {
    state: {
      todos: mockedTodos,
    },
    dispatch: jest.fn(),
  };

  jest
    .spyOn(AppContext, "useAppContext")
    // @ts-ignore
    .mockImplementation(() => mockContextValue);
});

describe("snapshot", () => {
  test("should return snapshot", () => {
    const tree = renderer.create(<TodoListItems id={todoId} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("handle events", () => {
  // afterEach(() => {
  //   jest.clearAllMocks();
  // });

  const mockClassListAddMethod = jest.fn();
  const setStateMock = jest.fn();
  const useStateMock = (initialValue: any) => [initialValue, setStateMock];
  const useRefMock = () => ({
    current: {
      classList: {
        add: mockClassListAddMethod,
      },
    },
  });
  jest.spyOn(React, "useRef").mockImplementation(useRefMock);
  // @ts-ignore
  jest.spyOn(React, "useState").mockImplementation(useStateMock);

  // test("should update todo status live by Enter keyboard", () => {
  //   const event = {
  //     key: "Enter",
  //     target: {
  //       value: "todo item edited",
  //     },
  //   };

  //   const expectedTodos = [
  //     {
  //       id: todoId,
  //       content: event.target.value,
  //       user_id: "firstUser",
  //       status: TodoStatus.All,
  //       created_date: "",
  //     },
  //     {
  //       id: "orislaj",
  //       content: "todo 2",
  //       user_id: "firstUser",
  //       status: TodoStatus.All,
  //       created_date: "",
  //     },
  //     {
  //       id: "jiwuro",
  //       content: "todo 3",
  //       user_id: "firstUser",
  //       status: TodoStatus.Active,
  //       created_date: "",
  //     },
  //     {
  //       id: "iejals",
  //       content: "todo 4",
  //       user_id: "firstUser",
  //       status: TodoStatus.Active,
  //       created_date: "",
  //     },
  //     {
  //       id: "jfuqos",
  //       content: "todo 5",
  //       user_id: "firstUser",
  //       status: TodoStatus.Completed,
  //       created_date: "",
  //     },
  //     {
  //       id: "ahsyel",
  //       content: "todo 6",
  //       user_id: "firstUser",
  //       status: TodoStatus.Completed,
  //       created_date: "",
  //     },
  //   ];
  //   const updateTodoContent = jest.spyOn(actions, "updateTodoContent");
  //   const wrapper = shallow(<TodoListItems id={todoId} />);
  //   const listItemWrapper = wrapper.find({ "data-testid": "list-item-wrapper" });
  //   listItemWrapper.simulate("dblclick");
  //   const inputEditTodoContent = listItemWrapper.find({ "data-testid": "live-edit-todo-content" });
  //   const inputEditTodoContentProps = inputEditTodoContent.props();

  //   // handle on change
  //   inputEditTodoContentProps.onChange(event);
  //   expect(setStateMock).toHaveBeenCalledWith(expectedTodos);

  //   // handle on Enter keydown
  //   inputEditTodoContentProps.onKeyDown(event);
  //   expect(setStateMock).toHaveBeenCalledWith(expectedTodos);
  //   expect(updateTodoContent).toHaveBeenCalledWith(todoId, event.target.value);
  // });

  // TODO: discard todo status live by onblur
  // test('should discard todo status live by onblur', () => {
  //
  // });
});
