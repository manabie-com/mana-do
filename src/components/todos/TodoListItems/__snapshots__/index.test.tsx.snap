// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should return snapshot 1`] = `
<article
  className="sc-bczRLJ cYyWkG list-item"
  data-testid="list-item-wrapper"
  onDoubleClick={[Function]}
  style={Object {}}
>
  <div
    className="text_and_checkbox"
  >
    <input
      checked={false}
      data-testid="toggle-all"
      id="hfuwyr"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="hfuwyr"
    >
      todo 1
    </label>
  </div>
  <div
    className="delete"
  >
    <button
      className="delete-btn"
      data-testid="delete-all-todo-item"
      onClick={[Function]}
    >
      <svg
        height="18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16.97 0l.708.707L9.546 8.84l8.132 8.132-.707.707-8.132-8.132-8.132 8.132L0 16.97l8.132-8.132L0 .707.707 0 8.84 8.132 16.971 0z"
          fill="#494C6B"
          fillRule="evenodd"
        />
      </svg>
    </button>
  </div>
</article>
`;
